### PR TITLE
Update source-gen APIs according to review

### PR DESCRIPTION
**Pull request by layomia**
Implements API approved in https://github.com/dotnet/runtime/issues/45448#issuecomment-892099426. These are cosmetic changes to APIs called by generated code to configure type metadata, to make them easier to extend with new features in future versions of the generator/STJ.

Should be ported to 6.0.

### CODE REVIEW

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:1008]: There are several overloads without `<TElement>` not shown in the approved API including `CreateQueueInfo()` and `CreateStackInfo()`. I assume these are for non-generic collection support.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoValuesOfT.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoValuesOfT.cs:10]: Add `readonly` modifier here and to the other related "InfoValue" types?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:514]: Why is `KeyInfo` and `ElementInfo` not a const like the other properties?

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (danmoseley) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:240]: this is a reminder to me to always use the tool to update these ref files, as with the churn here it's hard to see what actually changed 😄

**Comment by layomia** [comment]
I want to call out new commit https://github.com/dotnet/runtime/pull/59042/commits/4ce31e9f1083f55f5ae1f18264b9b5fc76ba9846 ("Rename fast-path func name and add src-gen/JsonNode interop support"). It implements the proposed API for tomorrow - https://github.com/dotnet/runtime/issues/45448#issuecomment-915449217.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/Serialization/NodeInteropTests.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/Serialization/NodeInteropTests.cs:12]: Thanks for adding\converting these node tests.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.cs:19]: much cleaner :)

**Review by steveharter — APPROVED** [review]
LGTM pending API approval especially around the SerializeHandler\Serialize changes.

**Comment by steveharter** [comment]
Possible Mono\WASM issues:
```
/datadisks/disk1/work/AD8D0967/p/build/wasm/WasmApp.Native.targets(293,5): error : Failed to compile /datadisks/disk1/work/AD8D0967/w/A24F0912/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Tests.dll.bc -> /datadisks/disk1/work/AD8D0967/w/A24F0912/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Tests.dll.o [/datadisks/disk1/work/AD8D0967/w/A24F0912/e/publish/ProxyProjectForAOTOnHelix.proj]
```

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:1011]: Yes those (without `<TElement>`) are for non-generic collection support. I believe the diff from the approved API is captured here - it mostly just encapsulates common arguments into the `JsonCollectionInfoValues<T>` type.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs
Comment (layomia) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:513]: I've made them `const` now. Didn't do that initially because they aren't used elsewhere.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:240]: 😅 I reverted extraneous diffs to make the actual changes clearer. Can follow up with a dedicated PR to clean up the ref file.

**Comment by layomia** [comment]
> LGTM pending API approval especially around the SerializeHandler\Serialize changes.

API has been approved, so this is good to merge when green - https://github.com/dotnet/runtime/issues/45448#issuecomment-919421589.

**Comment by eerhardt** [comment]
@lewing @radical @SamMonoRT - Have any of you seen the failure before? It appears to be an issue with WASM AOT.

https://helix.dot.net/api/2019-06-17/jobs/770ad140-31bb-4858-9396-d50162076efd/workitems/System.Text.Json.SourceGeneration.Roslyn3.11.Tests/console

```
  [39/47] xunit.runner.utility.netcoreapp10.dll.bc -> xunit.runner.utility.netcoreapp10.dll.o [took 20.326s]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : Failed to compile /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc -> /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.o [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error :  "/datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/bin/clang++" -target wasm32-unknown-emscripten -DEMSCRIPTEN -fno-inline-functions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=23 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=/datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -Oz -g -v -c /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc -o /tmp/tmpkKeaiw.tmp [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 5852582532b3eb3ea8da51a1e272d8d017bd36c9) [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : Target: wasm32-unknown-emscripten [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : Thread model: posix [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : InstalledDir: /datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/bin [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error :  (in-process) [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error :  "/datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/bin/clang++" -cc1 -triple wasm32-unknown-emscripten -emit-obj --mrelax-relocations -disable-free -main-file-name System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc -mrelocation-model static -mframe-pointer=none -fno-rounding-math -mconstructor-aliases -target-cpu generic -fvisibility hidden -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -v -fcoverage-compilation-dir=/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish -resource-dir /datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/lib/clang/13.0.0 -Oz -Werror=implicit-function-declaration -fdebug-compilation-dir=/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish -ferror-limit 19 -fgnuc-version=4.2.1 -fno-inline-functions -vectorize-slp -iwithsysroot/include/SDL -iwithsysroot/include/compat -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -o /tmp/tmpkKeaiw.tmp -x ir /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : clang -cc1 version 13.0.0 based upon LLVM 13.0.0git default target x86_64-unknown-linux-gnu [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
/datadisks/disk1/work/9D6C08B7/p/build/wasm/WasmApp.Native.targets(293,5): error : emcc: error: '/datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/bin/clang++ -target wasm32-unknown-emscripten -DEMSCRIPTEN -fno-inline-functions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=23 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=/datadisks/disk1/work/9D6C08B7/p/build/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -Oz -g -v -c /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc -o /tmp/tmpkKeaiw.tmp' failed (received SIGKILL (-9)) [took 433.767s] [/datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/publish/ProxyProjectForAOTOnHelix.proj]
XHarness artifacts: /datadisks/disk1/work/9D6C08B7/w/AA0A0963/uploads/xharness-output
```

**Comment by radical** [comment]
```
Failed to compile /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.bc -> /datadisks/disk1/work/9D6C08B7/w/AA0A0963/e/wasm_build/obj/wasm/System.Text.Json.SourceGeneration.Roslyn3.11.Tests.dll.o


... received SIGKILL (-9)
```

@lewing @radekdoulik Could this be due to running out of memory, or other resources?

**Comment by layomia** [comment]
Was chatting with @eerhardt about the AOT failure offline.. FWIW I wanted to note that the size of the test .dll in question here (`System.Text.Json.SourceGeneration.Roslyn4.0.Tests`) increased from 4.2 MB to 5.4 MB (+ ~28%) with these new changes. I expect the size to keep increasing as we continue sharing tests between src-gen/reflection `JsonSerializer` implementations.

**Comment by lewing** [comment]
cc @vargaz

**Comment by layomia** [comment]
In https://github.com/dotnet/runtime/pull/59042/commits/a6c3d7c72c87f6906b7a4e77abd90f6abcaf1e19 I disabled the src-gen tests in the WASM/AOT leg to unblock this PR so we can get the changes into 6.0. I opened https://github.com/dotnet/runtime/issues/59228 to track the disabled tests.

**Comment by MichalStrehovsky** [comment]
I thought RC1 is our go-live release and people can start shipping on it, including e.g. publishing NuGet packages with pregenerated code in it.

Will the NuGet packages using the RC1 S.T.Json source generator be usable on the RTM runtime?

I'm looking at a ```MissingMethodException: Missing method 'System.Text.Json.Serialization.Metadata.JsonTypeInfo`1<!!0[]> System.Text.Json.Serialization.Metadata.JsonMetadataServices.CreateArrayInfo(System.Text.Json.JsonSerializerOptions, System.Text.Json.Serialization.Metadata.JsonTypeInfo, System.Text.Json.Serialization.JsonNumberHandling, System.Action`2<System.Text.Json.Utf8JsonWriter, !!0[]>)'``` that appears to be caused by using an RC1 source generator to generate the code, but running on a post-RC1 runtime.

**Comment by danmoseley** [comment]
@MichalStrehovsky support for go-live releases like RC1 ends as soon as the next one goes out, RC2 in this case.
https://dotnet.microsoft.com/platform/support/policy/dotnet-core

**Pull request merged**
